### ISSUE TITLE
Use new `Store` interface from `express-rate-limit` v6

### ISSUE THREAD

**Issue by gamemaker1** [2do2go/rate-limit-mongo#40]
Hi @okv,

I recently [helped out with a Typescript rewrite](https://github.com/nfriedly/express-rate-limit/pull/263) for the [`express-rate-limit`](https://github.com/nfriedly/express-rate-limit) package, which also introduced [a couple of changes to the `Store` interface](https://github.com/nfriedly/express-rate-limit/blob/master/changelog.md?plain=1#L95-L100):

- Added the [`init`](https://github.com/nfriedly/express-rate-limit/blob/master/source/types.ts#L125-L131) method for stores to set themselves up using options passed to the middleware.
- Rename the [`incr`](https://github.com/nfriedly/express-rate-limit/blob/master/source/types.ts#L93-L99) method to [`increment`](https://github.com/nfriedly/express-rate-limit/blob/master/source/types.ts#L133-L140).
- Allow the `increment`, `decrement`, `resetKey` and `resetAll` [methods to return a promise](https://github.com/nfriedly/express-rate-limit/commit/ba782e10d36fac6eb3b1ad29febd2b697c08cd2e).
- Old stores [will automatically be promisified](https://github.com/nfriedly/express-rate-limit/commit/ba782e10d36fac6eb3b1ad29febd2b697c08cd2e) and used.

As [the external tests show](https://github.com/nfriedly/express-rate-limit/runs/4784820320?check_suite_focus=true), the store still works (without any changes) with v6 of `express-rate-limit`. However, implementing the newer `Store` interface allows for the following:

- The users do not need to pass [`expireTimeMs`](https://github.com/2do2go/rate-limit-mongo#configuration)/[`windowMs`](https://github.com/nfriedly/express-rate-limit#windowms) twice ([once to the middleware and once to the store](https://github.com/nfriedly/express-rate-limit/commit/16716b4d5568f77292967b65b2f87da462cf3294#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R120)) with the introduction of the [`init`](https://github.com/nfriedly/express-rate-limit/blob/master/source/types.ts#L125-L131) method.
- Since all functions [can now return Promises](https://github.com/nfriedly/express-rate-limit/commit/ba782e10d36fac6eb3b1ad29febd2b697c08cd2e), it removes the need for the numerous callbacks used in the code.

I would be happy to open/help out with a PR that implements the new `Store` interface.

Also, I see that #27 is still open and there is no `@types/rate-limit-mongo` package. Would you like this PR to also:

1. Rewrite the library in Typescript, *OR*
2. Add a type definition file (`index.d.ts` like what @gtmsingh has suggested in https://github.com/2do2go/rate-limit-mongo/issues/27#issuecomment-855444547) so that Typescript users don't encounter errors while using this library.

Please let me know your thoughts on the above.

Thanks,
Vedant

// cc @nfriedly

**Comment by okv** [comment]
Hey @gamemaker1, to be honest I'm not a big TypeScript fan. But everything except it sounds awesome :smiley:
I would keep the code in JS, but `index.d.ts` could be useful. Any assistance with reviewing #28 is highly appreciated.
I will try to take a closer look at all this next weekend.

**Comment by gamemaker1** [comment]
Great, I'll start working on a PR this weekend! I'll also take a look at #28.

**Comment by FxllenCode** [comment]
Is this issue stale? Not to necrobump but it would be nice to support v6 👍

**Comment by gamemaker1** [comment]
@FxllenCode yes, this issue is stale since I have not gotten the time to convert the store to by yet :(

Although ypu could send in a PR as per the discussion above.

**Comment by Sammaye** [comment]
I just quickly made this in a few hours, https://github.com/Sammaye/rate-limit-mongo it keeps to JS and uses the v6 interface, and it passes 60% of unit tests, the last 3 folders need merging into one since I only use a _getCollection() function now for all of it, also I have not updated the test framework version, I just got the tests to work with the new stuff.

But it is most of a recode.
